net/adoptopenjdk/bumblebench/collections/ArrayCopyBench.java

```diff
@@ -0,0 +1,121 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+package net.adoptopenjdk.bumblebench.collections;
+
+import net.adoptopenjdk.bumblebench.core.MicroBench;
+
+/**
+ * 
+ * 
+ * This benchmark simply tests the speed of copying arrays in either backward
+ * or forward direction. This direction can be specified by options.
+ * 
+ * The reported score is in terms of number of array copies done in one second.
+ * Length of array is fixed to 8192 but type can be changed using options.
+ * Also if not specified this benchmark will always test array copy for following types.
+ * rimitive Arrays(int, long, byte, character) and Object Arrays. This can be limited to
+ * specific type using options.
+ * 
+ * Also this benchmark in all iterations copies 4000 elements which can be changed
+ * through options.
+ * 
+ * */
+
+
+public final class ArrayCopyBench extends MicroBench {
+	private static final boolean BACKWARD_ARRAYCOPY = option("backwardArrayCopy", false);
+	private static final int ARRAY_LENGTH = option("arrayLength", 8192);
+	private static final int COPY_ELEMENTS = option("copyElements", 4000);
+	private static final int MAX_ITERATIONS_PER_LOOP = option("maxIterations", 10000000);
+	private static final boolean TESTINT = option("testInt", false);
+	private static final boolean TESTLONG = option("testLong", false);
+	private static final boolean TESTOBJ = option("testObj", false);
+	private static final boolean TESTCHAR = option("testChar", false);
+	private static final boolean TESTBYTE = option("testByte", false);
+	
+	private static int[] intArray = new int[ARRAY_LENGTH];
+	private static char[] charArray = new char[ARRAY_LENGTH];
+	private static long[] longArray = new long[ARRAY_LENGTH];
+	private static byte[] byteArray = new byte[ARRAY_LENGTH];
+	private static String[] objArray = new String[ARRAY_LENGTH];
+	private static int dstPoint;
+	private static int srcPoint;
+	static {
+		for (int i=0; i < ARRAY_LENGTH; i++) {
+			intArray[i] = i;
+			charArray[i] = (char)i;
+			longArray[i] = (long)i;
+			byteArray[i] = (byte)i;
+			objArray[i] = Integer.toString(i);
+		}
+		
+		if (BACKWARD_ARRAYCOPY) {
+			// Will be doing arraycopy in backward direction
+			srcPoint=0;
+			dstPoint=10;
+		}
+		else {
+			// Will be doing arraycopy in forward direction
+			srcPoint=10;
+			dstPoint=0;
+		}
+		
+	}
+	protected int maxIterationsPerLoop() {
+		return MAX_ITERATIONS_PER_LOOP;
+	}	
+	@Override
+	protected long doBatch(long numIterations) throws InterruptedException {
+		for (long loop = 0; loop < numIterations; loop++) {
+			if (TESTINT) {
+				startTimer();
+				System.arraycopy(intArray, srcPoint, intArray, dstPoint, COPY_ELEMENTS);
+				pauseTimer();
+			}
```

Comment:
I think we should not rely on the optimizer to version out these loops. Can we manually version out the `if` blocks here into separate loops since they are mutually exclusive?